spirv_msl.cpp

```diff
@@ -10699,6 +11735,97 @@ void CompilerMSL::OpCodePreprocessor::check_resource_write(uint32_t var_id)
 		uses_resource_write = true;
 }
 
+void CompilerMSL::access_chain_internal_append_index(std::string &expr, uint32_t base, const SPIRType *type,
+                                                     AccessChainFlags flags, bool &access_chain_is_arrayed,
+                                                     uint32_t index)
+{
+	bool index_is_literal = (flags & ACCESS_CHAIN_INDEX_IS_LITERAL_BIT) != 0;
+	bool register_expression_read = (flags & ACCESS_CHAIN_SKIP_REGISTER_EXPRESSION_READ_BIT) == 0;
```

Comment:
I think I figured out what the purpose of this function is. The case mostly happens when trying to OpLoad an array-of-array of inputs in TESE. This happens when TESC emits an array per-control point.

This function then handles the interleave-like operation which happens when trying to index into the loaded array-of-array with CompositeExtract.

This smells rather fishy, because input arrays should be unrolled already. Perhaps this is a special case which is only needed for array-of-array because I believe there are some missing cases for the current input unrolling code.

I'd rather fix that problem than adding a crude hack like this. The comment below also suggests this is a workaround.

Reply:
I'll make a separate issue out of this, and try to fix it in a cleaner way, then, this override should be possible to remove altogether.